### DISCUSSION TITLE
Crosstalk between Drivers

### DISCUSSION THREAD

**Discussion by mplaum** [dbeaver/dbeaver#20097] [General]
### Description

We have a vendor-supplied JDBC driver which we use to connect to their cloud data source. 

The driver is designed so that a configuration file (with endpoint details, credentials etc.) are co-located with the driver, and these have a specific file name - so they must be in separate folders. 

This works correctly when we connect to only one endpoint.

When connecting to separate endpoints, we have to create separate driver definitions for each one (in separate folders), since we can't define a separate home folder at the connection level.

Here's where the issue lies: If we open one connection using a specific driver, it opens correctly. Opening a second connection causes an error, since the driver being used for the previous connection is already loaded, and then causes an error when loading the new driver. 

The only way to open the second connection is to close DBeaver and reopen it, and select that connection first.

The error is:

`ClassCastException: attempting to castjar:file:/C:/Connections/Tenant_01/infor-compass-jdbc-2022.10 .jar!/com/infor/idl/shaded/javax/ws/rs/client/ClientBuilder.class to jar:file:/C:/Connections/Tenant_02/infor-compass-jdbc-2022.10 .jar!/com/infor/idl/shaded/javax/ws/rs/client/ClientBuilder.class`

So we can see that the driver from Tenant_01 is still loaded while we are trying to load the copy of the same driver in Tenant_02.

I have tried to resolve this by re-naming the .jar files (giving them unique names), but this does not correct the issue.

### DBeaver Version

Community Edition 23.0.1

### Operating System

Windows 10

### Database and driver

Database: Infor Data Lake
Driver: infor-compass-jdbc-2022.10 .jar

### Steps to reproduce

1. Create a JDBC driver definition for the first connection
2. Specify Class Name: com.infor.idl.jdbc.Driver
3. Specify URL Template: jdbc:infordatalake://{host}
4. Driver Libraries: Add file c:\connections\Tenant_01\infor-compass-jdbc-2022.10 .jar
5. Driver Libraries: Add folder c:\connections\Tenant_01 (this folder contains the tenant configuration file)
6. Create a JDBC driver definition for the second connection
7. Specify Class Name: com.infor.idl.jdbc.Driver
8. Specify URL Template: jdbc:infordatalake://{host}
9. Driver Libraries: Add file c:\connections\Tenant_02\infor-compass-jdbc-2022.10 .jar
10. Driver Libraries: Add folder c:\connections\Tenant_02 (this folder contains the tenant configuration file)
11. Create connection using the driver from Tenant_01
12. Specify the hostname for Tenant_01
13. Create connection using the driver from Tenant_02
14. Specify the hostname for Tenant_02
15. Open connection for Tenant_01
16. Verify that the connection opens correctly
17. Open connection for Tenant_02
16. Observe the error
17. Close the application
18. Open the application
19. Open connection for Tenant_02
20. Verify that the connection opens correctly
21. Open connection for Tenant_01
22. Observe the error

**Comment by ShadelessFox** [comment]
Hello @mplaum,

Instead of putting the jar (`infor-compass-jdbc-2022.10.jar`) directly into the driver, try adding it globally in the preferences:

![image](https://user-images.githubusercontent.com/35821147/228887961-8706dc2c-6550-4254-a59d-db1a98a5ff01.png)

Please let me know if it helps.

**Comment by mplaum** [comment]
Thank you for the great suggestion, @ShadelessFox!

I gave that a test - for each of the driver definitions I added the folder with the configuration file (Infor Compass JDBC Driver.ionapi).

The error that I get now is this:

```
JDBC Error 100

Unable to connect. The Infor Compass JDBC Driver.ionapi file is not found. The file must be located in the folder containing the Compass JDBC driver JAR file.

2023-03-30T15:53:03.539440900Z
```

Unfortunately the driver file name is not configurable, and is the same for all of the tenants, so they have to be in separate folders.

**Comment by ShadelessFox** [comment]
Can you please send the log files when the error occurs?

You can find instructions here: https://dbeaver.com/docs/wiki/Log-files/. Thank you.

**Comment by ShadelessFox** [comment]
Unfortunately, GitHub can't attach files when you answer by email, so I don't see the files you sent. Can you please go to GitHub and attach them there?

**Comment by mplaum** [comment]
Log file attached. 
[DBeaver_InforDataFabric_Error_20230503.log](https://github.com/dbeaver/dbeaver/files/11398145/DBeaver_InforDataFabric_Error_20230503.log)

Thank you!

**Comment by ShadelessFox** [comment]
It looks like you still have two separate jars added for two drivers, respectively.

Keep just a single jar in a global classpath and remove them from drivers' files.

**Comment by mplaum** [comment]
I will give that a try! Thank you.

**Comment by mplaum** [comment]
Unfortunately that didn't work - The driver requires the connection detail file to be located in the same folder as the driver. Since the connection detail file has a specific file name (Infor Compass JDBC Driver.ionapi), and we have three tenants that we are using this with. we can't have all of them in the same folder as the driver since they have the same name.

**Comment by ShadelessFox** [comment]
How is the driver supposed to be used? I'm not following. It forces the user to use two separate jar files, but it seems impossible.

**Comment by mplaum** [comment]
I can give an example of how we use it with our ODBC/JDBC bridge driver - I hope it helps:

- We set up a separate folder for each tenant that we want to connect to
- Each folder contains the "Infor Compass JDBC Driver.ionapi" file for its respective tenant
- We place a copy of the driver .jar file in each folder
- Using the ODBC driver manager, we create a System DSN for each endpoint, referencing the respective driver that is co-located with its .ionapi file for the respective tenant

We were trying to use the same method with DBeaver - the issue that is causing us grief is the requirement that the .ionapi file needs to be in the same folder as the driver - that's how the driver is coded to locate that file.

**Comment by ShadelessFox** [comment]
The issue is not related to DBeaver.

Please look at the JDBC driver's docs on how to use the driver in this specific scenario.

**Comment by mplaum** [comment]
Just an update - I found a way to make it work. The Infor Data Fabric allows the .ionAPI file contents to be added to the connection string. These need to be URLEncoded first.

[From the Infor Data Fabric User Guide](https://docs.infor.com/inforos/2021.x/en-us/datafabrug/default.html?helpcontent=fmu1631199547459.html):
```
1. Open the SQL query tool.
2. Select the option to set up a new driver.
3. Specify a driver name, such as Compass JDBC Driver.
4. Specify the URL: jdbc:infordatalake://TENANT_TST?ionApiCredentials=...
     a. Using a URL Encoding method, URL-encode the contents of the .ionapi file.
        The encoded result must start with a “%7B” representing the opening left curly brace “{“ and end with “%7D” which is closing bracket.
     b. Take the encoded credentials and add them to the end of the URL connection string.
         Example: jdbc:infordatalake://TENANT_TST?ionApiCredentials=YourEncoded Credentials
5. Set the classname to com.infor.idl.jdbc.Driver.
6. Specify the library path to the JDBC driver - e.g. c:\compass\infor-compass-jdbc-2020-09.jar
7. Save the driver.
8. Set up a connection to the Data Lake and select the Compass JDBC driver.
9. Connect to Infor Data Lake.
```
